File: services/client/src/protocol/connection.go
package protocol

import (
	"io"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

/*
 * Se encarga de enviar un mensaje completo utilizando el protocolo definido
 *
 * Primero se construye y envía el header de tamaño fijo
 * Luego envía el payload solamente cuando el mensaje contiene datos
 *
 * SendAll garantiza el envio de todos los bytes
 */
func SendMessage(writer io.Writer, messageType MessageType, payload []byte,) error {

	message := Message{
		Header: Header{Type: messageType, PayloadLength: uint32(len(payload))},
		Payload: payload,
	}

	if err := ValidateMessage(message); err != nil {
		return err
	}

	// Serializo el header
	headerBytes, err := EncodeHeader(message.Header)
	if err != nil {
		return err
	}

	// Envío todos los bytes del header
	if err := safe_socket.SendAll(writer, headerBytes); err != nil {
		return err
	}

	// Algunos tipos pueden no tener payload
	if len(payload) == 0 {
		return nil
	}

	// Envío todos los bytes correspondientes al payload.
	if err := safe_socket.SendAll(writer, payload); err != nil {
		return err
	}

	return nil
}

/*
 * Se encargar de recibir un mensaje completo utilizando el protocolo definido
 *
 * Primero recibe los HeaderSize bytes del header 
 * Deserializa, y a partir de eso sabe cuantos bytes debe recibir del payload
 *
 * SendAll garantiza el envio de todos los bytes
 */
func ReceiveMessage(reader io.Reader) (Message, error) {

	headerBytes, err := safe_socket.RecvAll(reader, HeaderSize)
	if err != nil {
		return Message{}, err
	}

	// Deserializo el header para conocer el tipo y tamaño del payload.
	header, err := DecodeHeader(headerBytes)
	if err != nil {
		return Message{}, err
	}

	payload := []byte{}

	// Recibo exactamente la cantidad de bytes indicada en el header
	if header.PayloadLength > 0 {
		payload, err = safe_socket.RecvAll(
			reader,
			int(header.PayloadLength),
		)

		if err != nil {
			return Message{}, err
		}
	}

	message := Message{
		Header:  header,
		Payload: payload,
	}

	if err := ValidateMessage(message); err != nil {
		return Message{}, err
	}

	return message, nil
}

File: services/server/src/protocol/connection.py
import socket

import safe_socket

from .codec import decode_header, encode_header, validate_message
from .message import HEADER_SIZE, Header, Message, MessageType


"""
Primero recibe el header de tamaño fijo y luego utiliza el tamaño
indicado en el header para recibir exactamente los bytes del payload
"""
def receive_message(socket: socket.socket) -> Message:
    header_bytes = safe_socket.recv_all(
        socket,
        HEADER_SIZE,
    )

    if not header_bytes:
        raise ConnectionError("socket closed before receiving message header")

    # Deserializo el header para conocer el tipo y tamaño del payload
    header = decode_header(header_bytes)

    payload = b""

    if header.payload_length > 0:
        payload = safe_socket.recv_all(
            socket,
            header.payload_length,
        )

        if len(payload) != header.payload_length:
            raise ConnectionError("socket closed before receiving complete payload")

    message = Message(header=header,payload=payload)

    validate_message(message)

    return message


"""
Envia un mensaje completo utilizando el protocolo definido.

Primero serializa y envia el header y luego envia el payload
solamente cuando el mensaje contiene datos.
"""
def send_message(socket: socket.socket,message_type: MessageType,payload: bytes = b"") -> None:
    header = Header(
        message_type=message_type,
        payload_length=len(payload),
    )

    message = Message(
        header=header,
        payload=payload,
    )

    validate_message(message)

    # Serializo y envio el header
    header_bytes = encode_header(header)

    safe_socket.send_all(
        socket,
        header_bytes,
    )

    # Algunos tipos de mensajes pueden no tener payload
    if payload:
        safe_socket.send_all(
            socket,
            payload,
        )